### PR TITLE
Pass Task Properties to Task Parameters

### PR DESCRIPTION
**Pull request by splatteredbits**
It would be nice if Whiskey could pass task properties from YAML to a task function's named parameters.

### CODE REVIEW

**Comment by splatteredbits** [comment]
@whsdevops Finally ready for a review on this.

File: Assembly/Whiskey/Tasks/ParameterValueFromVariableAttribute.cs [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+{
+    public class ParameterValueFromVariableAttribute : Attribute
+    {
+        public ParameterValueFromVariableAttribute(string name)
+        {
+            VariableName = name;
+        }
+
+        public string VariableName { get; set; }
+
```
Comment (pnstickne) [inline Assembly/Whiskey/Tasks/ParameterValueFromVariableAttribute.cs]: Extra newline. Perhaps read-only property?

File: Test/Get-TaskParameter.Tests.ps1 [hunk 2]
```diff
@@ -0,0 +1,611 @@
+
+#Requires -Version 5.1
+Set-StrictMode -Version 'Latest'
+
+& (Join-Path -Path $PSScriptRoot -ChildPath 'Initialize-WhiskeyTest.ps1' -Resolve)
+
+[Whiskey.Context]$context = $null
+$global:taskCalled = $false
+$global:taskParameters = $false
```
Comment (DecoyJoe) [inline Test/Get-TaskParameter.Tests.ps1]: Nit: Should be `$global:taskParameters = $null`

File: Test/Get-TaskParameter.Tests.ps1 [hunk 3]
```diff
@@ -0,0 +1,611 @@
...
+        {
+            Remove-Item -Path $path
+        }
+    }
+}
+
+function ThenPipelineSucceeded
+{
+    $Global:Error | Should -BeNullOrEmpty
+    $threwException | Should -Be $false
```
Comment (DecoyJoe) [inline Test/Get-TaskParameter.Tests.ps1]: Nit: `Should -Be $false` can be updated to `Should -BeFalse`

File: Test/Get-TaskParameter.Tests.ps1 [hunk 4]
```diff
@@ -0,0 +1,611 @@
...
+    $taskParameters | Should -BeNullOrEmpty
+}
+
+function ThenThrewException
+{
+    param(
+        $Pattern
+    )
+
+    $threwException | Should -Be $true
```
Comment (DecoyJoe) [inline Test/Get-TaskParameter.Tests.ps1]: Nit: `Should -Be $true` can be updated to `Should -BeTrue`

File: Test/Get-TaskParameter.Tests.ps1 [hunk 5]
```diff
@@ -0,0 +1,611 @@
...
+        [string]
+        $Name,
+
+        [hashtable]
+        $Parameter
+    )
+
+    $script:context = New-WhiskeyTestContext -ForDeveloper
+    $context.PipelineName = 'Build';
+    $context.TaskName = $null;
```
Comment (DecoyJoe) [inline Test/Get-TaskParameter.Tests.ps1]: Nit: Unnecessary semicolons.

File: Test/Get-TaskParameter.Tests.ps1 [hunk 6]
```diff
@@ -0,0 +1,611 @@
...
+                [Whiskey.Tasks.ValidatePath()]
+                [string]
+                $Path
+            )
+            $global:taskCalled = $true
+            $global:taskParameters = $PSBoundParameters
+        }
+        Init
+        GivenFile 'abc.yml' 
+        WhenRunningTask 'Task' -Parameter @{ 'Path' = '*.yml' } -ErrorAction SilentlyContinue
```
Comment (DecoyJoe) [inline Test/Get-TaskParameter.Tests.ps1]: How does this test work? I only see one `*.yml` file being created.

File: Test/Get-TaskParameter.Tests.ps1 [hunk 7]
```diff
@@ -0,0 +1,611 @@
...
+        }
+        Init
+        GivenFile 'abc.yml'
+        WhenRunningTask 'Task' -Parameter @{ 'Path' = 'abc.yml' } -ErrorAction SilentlyContinue
+        ThenTaskNotCalled
+        ThenThrewException 'must\ be\ a\ directory'
+    }
+}
+
+Describe ('Get-TaskParameter.when path should be a file and passed multiple patht and one isn''t') {
```
Comment (DecoyJoe) [inline Test/Get-TaskParameter.Tests.ps1]: Spelling and looks like an incomplete sentence.

File: whiskey.yml [hunk 8]
```diff
@@ -37,8 +38,14 @@ Build:
     Path:
     - Test\about_Whiskey_Tasks.Tests.ps1
 
+- Pester4:
+    OnlyOnPlatform: Linux
```
[the remark this answers is not in the document]
Comment (splatteredbits) [reply whiskey.yml]: Yes. Improved on another branch, but even with that fix, they don't like to run in parallel. :(

File: Whiskey/en-US/about_Whiskey_Writing_Tasks.help.txt [hunk 9]
```diff
@@ -46,14 +29,77 @@ LONG DESCRIPTION
...
+                [hasthable]
+                $TaskParameter
+            )
+
+            Set-StrictMode -Version 'Latest'
+        }
+        
+    Named parameters are never passed in the `TaskParameter` hashtable.
+
+    If you want the value of your parameter to come from a Whiskey variable, decorate the parameter with Whiskey's `Whiskey.Tasks.ParameterValueFromVariable` attribute, passing the variable expression (method/property names allowed):
```
[the remark this answers is not in the document]
Comment (splatteredbits) [reply Whiskey/en-US/about_Whiskey_Writing_Tasks.help.txt]: It's for making it easier to port existing functions to work as Whiskey tasks and as stand-alone functions. This takes precedence over any value provided by the user.

File: Whiskey/en-US/about_Whiskey_Writing_Tasks.help.txt [hunk 10]
```diff
@@ -46,14 +29,77 @@ LONG DESCRIPTION
             - One
             - Two
 
-    Everything under your task in the whiskey.yml file is passed to your function to the `$TaskParameter` hashtable parameter. For example, given the whiskey.yml above, your task would receive this hashtable:
+    To allow users to pass values to your task, create parameters on your function. Whiskey will map the parameter names to the property names from the build YAML.
```
[the remark this answers is not in the document]
Comment (splatteredbits) [reply Whiskey/en-US/about_Whiskey_Writing_Tasks.help.txt]: The universe of types that won't work is pretty large. I'll update the docs to mention that PowerShell has to be able to implicitly cast strings to the type of the parameter.

File: Whiskey/en-US/about_Whiskey_Writing_Tasks.help.txt [hunk 11]
```diff
@@ -66,7 +112,79 @@ LONG DESCRIPTION
...
+            param(
+                [Whiskey.Tasks.ValidatePath()]
+                [string]
+                $Path
+            )
+
+            Set-StrictMode -Version 'Latest'
+        }
+
+    By default, the property can be empty and any values from the YAML file *must* exist. If you want to require that the user provide a value (and fail a build if it doesn't), use the `Mandatory` property:
```
[the remark this answers is not in the document]
Comment (splatteredbits) [reply Whiskey/en-US/about_Whiskey_Writing_Tasks.help.txt]: They mean the same thing, but in order to find out if a parameter is mandatory, Whiskey would have to parse all the potential parameter sets in a function, identify which one is being used, and then determine if it is mandatory. It is simpler to have a `Mandatory` property on Whiskey's attribute.

[the remark this answers is not in the document]
Comment (splatteredbits) [reply Whiskey/en-US/about_Whiskey_Writing_Tasks.help.txt]: Documentation slightly re-worded.

File: Test/Get-TaskParameter.Tests.ps1 (hunk 6 above)
Comment (splatteredbits) [reply Test/Get-TaskParameter.Tests.ps1]: The tests create a default whiskey.yml file. Should be clearer now.

**Pull request merged**